python udf over rpc

zhaojuanmao opened the pull request:
Summary:
This diff is to support python user defined function over rpc for #23110, work flow is like this:
1. pickle python udf
2. pass pickle to C++
3. C++ pass over rpc from client to server
4. server call runPythonUDF() python function to unpickle and run python udf and pickle the udf result using python embedder
6. pass back serialized result from server to client
7. client call loadPythonUDFResult() python function to unpickle result
7. return it to python

right now, put rpc_sync_builtin() and rpc_async_builtin() as temporary interfaces for builtin operator remote calls, they accept qualified name string, this interface can execute builtin operators in C++ land.

rpc_sync() and rpc_async() accept python callables only right now, it could be user define python functions or builtin operator python functions, the python functions will be executed in python land.

once we can resolve builtin operator python callables to qualified name string, we can merge rpc_sync_builtin() into rpc_sync() then

Differential Revision: D16390764

xush6528 commented on torch/distributed/rpc.py:
```diff
@@ -0,0 +1,127 @@
+        of ``func`` on ``args`` and ``kwargs`` can be retrieved from the Future
+        object.
+    """
+    qualified_name = torch.jit._find_builtin(func)
```
we'd better assert here.
```Python
assert _agent is not None, "init_rpc(..) has not been called to setup rpc_agent yet."
```
Otherwise, the error message could look like
```Bash
Traceback (most recent call last):
  File "/usr/local/fbcode/platform007/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/local/fbcode/platform007/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/data/users/shihaoxu/fbsource/fbcode/buck-out/dev/gen/caffe2/torch/fb/modelparallel/prototype/pytorch/tests/test_rpc#binary,link
-tree/caffe2/torch/fb/modelparallel/prototype/pytorch/rpc.py", line 1057, in _peer_ping
    _low_level_rpc_no_result(_rpc_ping, inputs=[], worker_id=i)
  File "/data/users/shihaoxu/fbsource/fbcode/buck-out/dev/gen/caffe2/torch/fb/modelparallel/prototype/pytorch/tests/test_rpc#binary,link
-tree/caffe2/torch/fb/modelparallel/prototype/pytorch/rpc.py", line 484, in _low_level_rpc_no_result
    _call_rpc(RPCRequest(func, inputs, kwargs), worker_id)
  File "/data/users/shihaoxu/fbsource/fbcode/buck-out/dev/gen/caffe2/torch/fb/modelparallel/prototype/pytorch/tests/test_rpc#binary,link
-tree/caffe2/torch/fb/modelparallel/prototype/pytorch/rpc.py", line 472, in _call_rpc
    dst_name=WorkerCtx.instance().name_for_id(worker_id),
  File "/data/users/shihaoxu/fbsource/fbcode/buck-out/dev/gen/caffe2/torch/fb/modelparallel/prototype/pytorch/tests/test_rpc#binary,link
-tree/caffe2/torch/fb/modelparallel/prototype/pytorch/comm.py", line 131, in send_callable
    dist.rpc_async(dst_name, _run_request, obj)
  File "/data/users/shihaoxu/fbsource/fbcode/buck-out/dev/gen/caffe2/torch/fb/modelparallel/prototype/pytorch/tests/test_rpc#binary,link
-tree/torch/distributed/rpc.py", line 95, in rpc_async
    return invoke_rpc_python_udf(_agent, to, serialize(PythonUDF(func, args, kwargs)))
TypeError: invoke_rpc_python_udf(): incompatible function arguments. The following argument types are supported:
    1. (arg0: torch.distributed.RpcAgent, arg1: str, arg2: str) -> torch.distributed.FutureMessage

Invoked with: None, 'w:0', b'\x80\x03ctorch.distributed.internal_rpc_utils\nPythonUDF\nq\x00ccaffe2.torch.fb.modelparallel.prototype.pyt
orch.comm\n_run_request\nq\x01ccaffe2.torch.fb.modelparallel.prototype.pytorch.rpc\nRPCRequest\nq\x02)\x81q\x03}q\x04(X\x04\x00\x00\x00f
uncq\x05ccaffe2.torch.fb.modelparallel.prototype.pytorch.rpc\n_rpc_ping\nq\x06X\x06\x00\x00\x00callerq\x07X\x03\x00\x00\x00w:1q\x08X\x06
\x00\x00\x00outputq\tNX\x06\x00\x00\x00inputsq\n]q\x0bX\x06\x00\x00\x00kwargsq\x0cNX\x0c\x00\x00\x00client_scopeq\rccaffe2.torch.fb.mode
lparallel.prototype.pytorch.rpc\nRRefId\nq\x0eK\x01K\x00\x86q\x0f\x81q\x10ub}q\x11\x87q\x12\x81q\x13.'
Traceback (most recent call last):
```

zhaojuanmao replied:
in which case _agent could be None? maybe we should fix that first

zhaojuanmao replied:
in which case _agent is None? maybe we should fix that first

zhaojuanmao replied:
in which case _agent will be None? maybe we should fix that first

xush6528 commented on torch/csrc/distributed/rpc/PythonRpcHandler.cpp:
```diff
@@ -0,0 +1,89 @@
+PyObject* PythonRpcHandler::loadResultFunction_ = nullptr;
+
+void PythonRpcHandler::init() {
+  Py_Initialize();
```
I think we need to discuss more about this line.

It looks to me that our user starts with launching a Python interpreter (a C program) to interpret a piece of Python code. This Python interpreter, loads a pybind11 module, `torch.distributed`, that bundles several C++ classes.
We use this special binding module, backed by C++ classes, to instantiate a `RpcAgent` C++ instance, and assigned it to a variable `torch.distributed.rpc._agent`.
While, in the constructor of `RpcAgent`, `PythonRpcHandler::init()` is called and creates a **second** Python interpreter ?? (If I understand the [doc](https://docs.python.org/3/c-api/init.html#c.Py_Initialize) correctly)
This **second** Python interpreter imports the module, `torch.distributed.internal_rpc_utils`, and use the utilities in it. 
Although it can successfully run the function sent by the client and return correct results back to the client.
It's impossible for the client to change the global state in the server, which is hold by the **first** Python interpreter.

---
There is a requirement in our RRef prototype implementation that we need to allow an RPC client (an RRef user), to dictate the RPC server (RRef owner) to change it's global state. 
The state is stored in a module's global variable, which lives in the original Python Interpreter.

For example, 
1) A client could ask the server to fetch an RRef instance that matches the `ref_id` in the message, from the server's global RRef registry.
2) A client could send termination signal message to a server, the server, on receiving the signal, should set it's status to `termination_ongoing=True`, so that it will delete the remote references it holds and notifies the owners about reducing ref counts before it goes away.

xush6528 commented on torch/csrc/distributed/rpc/PythonRpcHandler.cpp:
```diff
@@ -0,0 +1,89 @@
+PyObject* PythonRpcHandler::loadResultFunction_ = nullptr;
+
+void PythonRpcHandler::init() {
+  Py_Initialize();
```
I think we need to discuss more about this line.

It looks to me that our user starts with launching a Python interpreter (a C program) to interpret a piece of Python code. This Python interpreter, loads a pybind11 module, `torch.distributed`, that bundles several C++ classes.
We use this special binding module, backed by C++ classes, to instantiate a `RpcAgent` C++ instance, and assigned it to a variable `torch.distributed.rpc._agent`.
While, in the constructor of `RpcAgent`, `PythonRpcHandler::init()` is called and creates a **second** Python interpreter ?? (If I understand the [doc](https://docs.python.org/3/c-api/init.html#c.Py_Initialize) correctly)
This **second** Python interpreter imports the module, `torch.distributed.internal_rpc_utils`, and use the utilities in it. 
Although it can successfully run the function sent by the client and return correct results back to the client.
It's impossible for the client to change the global state in the server, which is hold by the **first** Python interpreter.

---
There is a requirement in our RRef prototype implementation that we need to allow an RPC client (an RRef user), to dictate the RPC server (RRef owner) to change it's global state. 
The state is stored in a module's global variable, which lives in the original Python Interpreter.

For example, 
1) A client could ask the server to fetch an RRef instance that matches the `ref_id` in the message, from the server's global RRef registry.
2) A client could send termination signal message to a server, the server, on receiving the signal, should set it's status to `termination_ongoing=True`, so that it will delete the remote references it holds and notifies the owners about reducing ref counts before it goes away.
---
I read this doc, https://docs.python.org/3/extending/index.html#extending-index

It looks to me there are 2 ways for C/C++ to interact with Python.

1) > Write modules in C or C++ to extend the Python interpreter with new modules.
2) > Sometimes, rather than creating an extension that runs inside the Python interpreter as the main application, it is desirable to instead embed the CPython runtime inside a larger application.

Looks to me `Py_Initialize()` belongs to the interface for the seconds case.

zhaojuanmao replied:
I thought we only call Py_Initialize() once to import interpreter, as init() only is called once...? although we can enforce once as well, but I do not understand why it is called twice?

also could you please share what kind of error you encountered for the integration? we can sync up offline

zhaojuanmao replied:
I thought Py_Initialize() will only be called once, as init() is called only once...? although I can enforce it to be called only once explicitly as well, but I do not understand how it  could be called twice to import interpreter? we can sync up offline about this

xush6528 replied:
@zhaojuanmao 
NVM. There is a call to create a new Interpreter.
https://docs.python.org/3/c-api/init.html#c.Py_NewInterpreter

So I think `Py_Initialize` ensures singleton and initialize only once.

xush6528 commented on torch/csrc/distributed/rpc/PythonRpcHandler.cpp:
```diff
@@ -0,0 +1,89 @@
+  Py_DECREF(module_);
+  Py_DECREF(runUDFFunction_);
+  Py_DECREF(loadResultFunction_);
+  Py_Finalize();
```
This "kill"s Python interpreter.

https://docs.python.org/3/c-api/init.html#c.Py_FinalizeEx

zhaojuanmao replied:
I added a flag to ensure we will call Py_Finalize() only if Py_initialize() is called in embedder

xush6528 commented on torch/csrc/distributed/rpc/ProcessGroupAgent.cpp:
```diff
@@ -0,0 +1,213 @@
+  sendThread_.join();
+  listenerThread_.join();
+
+  PythonRpcHandler::cleanUp();
```
Can we move this into the destructor of ProcessGroupAgent.

ezyang replied:
More generally, the lifetime of PythonRpcHandler's "static" variables actually seems directly tied to `ProcessGroupAgent` lifetime. If this is true, can we just make PythonRpcHandler a member of ProcessGroupAgent?

zhaojuanmao replied:
that was initial thought, but turn out we need to call interface of PythonRpcHandler when we get result from future. At that point, future does not have processGroupAgent.

xush6528 commented on torch/csrc/distributed/rpc/PythonRpcHandler.cpp:
```diff
@@ -0,0 +1,89 @@
+PyObject* PythonRpcHandler::loadResultFunction_ = nullptr;
+
+void PythonRpcHandler::init() {
+  Py_Initialize();
```
We could move this line, `Py_Initialize();`.

As documentation says,
> This initializes the table of loaded modules (sys.modules), and creates the fundamental modules builtins, __main__ and sys. It also initializes the module search path (sys.path). It does not set sys.argv;

Since the code has run to here, the interpreter system-level modules must have already been initialized, so no need to call initialize again.

zhaojuanmao replied:
we still need this if the application is C++ application and no python environment. added to check Py_IsInitialized() or not first, to ensure Py_Initialize() is called only once

xush6528 commented on test/test_rpc.py:
```diff
@@ -0,0 +1,180 @@
+    def test_py_modify_global_var(self):
+        n = self.rank + 1
+        dstRank = n % self.world_size
+        ret = dist.rpc_sync('worker%d' % dstRank, modify_global_var, args=(True,))
```
This could be flaky, in worst case, a worker could sent out it's request, but hasn't received a request yet. In that case, the global var it owns is not changed.

This is the code to simulate the worst case.
```Python
        dstRank = n % self.world_size
        import time; time.sleep(self.rank)
        ret = dist.rpc_sync('worker%d' % dstRank, modify_global_var, args=(True,))
        self.assertEqual(global_var, True)
```


Fix:
```Python
        dstRank = n % self.world_size
        ret = dist.rpc_sync('worker%d' % dstRank, modify_global_var, args=(True,))
        dist.barrier()
        self.assertEqual(global_var, True)
```

zhaojuanmao replied:
good catch!

mrshenli replied:
Thinking more about this. It seems neither `dist.barrer()` nor the existing `sync_rpc` can guarantee a fix on this problem. Because they only enforces all sends are done, but there is no guarantee that all received messages are processed.

This can be fixed when we switch to `ThreadPool` for recv and let `sync_rpc` block until both send and recv are done. But the question is what should we wait for if received messages trigger subsequent sends (and those sends can trigger additional sends). 

I am thinking about the just doing following:
1. block send queue, and wait till all sends are done.
2. block recv queue, and wait till all recvs are done.

This means that we at least wait until existing sends are done and they are processed on the callee side (sufficient for this test). And we are not going to wait for additional send/recv triggered by existing sends. Does this make sense? @xush6528 @zhaojuanmao 

(this is not a change request for this PR, just discussion)

mrshenli replied:
In terms of changes, if we all agree that `dist.barrier` does not guarantee to fix the problem, shall we disable the test for now?

zhaojuanmao replied:
sounds good, let me disable the test for now

xush6528 commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -0,0 +1,65 @@
+      return torch::jit::createPyObjectForStack(std::move(stack));
+    }
+    case MessageType::PYTHON_RET: {
+      return PythonRpcHandler::loadPythonUDFResult(message);
```
Can we raise if the object is of `ExceptionBase` type?
Just like what Python `concurrent.futures.Future.results()` does.

Ref
https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.result



Otherwise, I have to unwrap the exception myself.
```Python
    fut = dist.rpc(func, inputs, kwargs, async_call=True)
    val = fut.wait()
    if isinstance(val, BaseException):
        raise val
    return val
```

zhaojuanmao replied:
will try to add one

zhaojuanmao replied:
@xush6528 I checked codes again, I think normal exception will be propagated and raised to user by calling fut.wait(), in this case "val = fut.wait()" will fail.  

in which case "val = fut.wait()" succeed and val is exception object? this looks like it is not system exception, but user anticipated exception?

xush6528 replied:
@zhaojuanmao 


```Python
# rpc.py

class RPCException(Exception):
    pass
```

```Python
# user_code.py

def raise_func():
    raise Exception("This is user exception")

fut = dist.rpc(to="worker1", raise_func, async_call=True)

try:
    return fut.wait()
except dist.RPCException as e:
    logger.error("RPC system failed.")
except BaseException:
    logger.error("User func failed to finish.")
```

zhaojuanmao replied:
if user call raise exception on server side, server will crash, so we need remote exception handling and return remote exception as a new type of message and raise properly on client side. this is needed for both builitin operator and python UDF, I think it is in follow up diff list

mrshenli commented on torch/csrc/distributed/rpc/Deserializer.cpp:
```diff
@@ -0,0 +1,23 @@
+#include <torch/csrc/distributed/rpc/Deserializer.h>
```
It seems one of my rebase messed up files. This file should not be here.

Same for `Serializer`, `Request`, `Response`.

mrshenli commented on test/test_rpc.py:
```diff
@@ -0,0 +1,181 @@
+        n = self.rank + 1
+        dstRank = n % self.world_size
+        ret = dist.rpc_sync('worker%d' % dstRank, my_class.my_class_method, args=(n, n + 1))
+        self.assertEqual(ret, (n + n + 1))
```
Is there any reason for hard code the result instead of using the following?

```python
args = (n, n+1)
ret = dist.rpc_sync('worker%d' % dstRank, my_class.my_class_method, args=args)
self.assertEqual(ret, my_class.my_class_method(*args))
```

zhaojuanmao replied:
I will change that

ezyang commented:
I got tagged as a reviewer; what would you like me to review in particular?

zhaojuanmao commented:
@ezyang that would be great if you can give some feedback about python embedding API, general code structure, thanks!

pritamdamania87 commented on torch/distributed/rpc.py:
```diff
@@ -94,13 +97,16 @@ def rpc_async(to, func, args=None, kwargs=None):
     kwargs = kwargs if kwargs else {}
-    return invoke_rpc(_agent, to, qualified_name, *args, **kwargs)
+
+    if qualified_name is not None:
```
What if our `func` name for python collides with a builtin op name?

zhaojuanmao replied:
builitin op start with torch., it should not collide with user defined function?

mrshenli commented:
Hi @zhaojuanmao, can you rebase this PR to show the latest changes? thanks!

mrshenli commented on torch/csrc/distributed/rpc/PythonRpcHandler.cpp:
```diff
@@ -0,0 +1,95 @@
+    throw std::runtime_error("python function args is NULL");
+    PyErr_Print();
+  }
+  auto pres = PyEval_CallObject(runUDFFunction_, pargs);
```
If the UDF makes some IO calls (e.g., another sync RPC), it will block one thread on the callee side (pointed out by @xush6528 and @pietern). In the short term, we can bump up the number of recv threads on `RpcAgent`, but that does not completely solve the problem. Consider two workers A and B. A makes an RPC call to B (ping), in that RPC call B makes another RPC back to A (pong), and A-B do nested ping-pong 100 times before exiting. We don't want this to block 100 threads on both sides. To solve this problem, we might need to expose a responder/FutureResult concept to Python land. For example, the UDF can take a `dist.FutureResult()` argument. The 100 nested ping-pong can be added as callbacks to asynchronous RPC results (`FutureMessage` objects). Then the last pong on B calls `futureResult.markCompleted(value)`, which answers the very first RPC from A.

This of course does NOT need to be included in this PR, and I think itself is complex enough and would require some design discussion. Does this make sense? @zhaojuanmao @pritamdamania87 @pietern @xush6528 @satgera

zhaojuanmao replied:
@mrshenli that makes sense, I will follow up on this design and work on it in follow up PR

ezyang commented:
Yes, a rebase would be much appreciated.

zhaojuanmao commented on torch/csrc/distributed/rpc/PythonRpcHandler.cpp:
```diff
@@ -0,0 +1,95 @@
+  module_ = PyImport_ImportModule("torch.distributed.internal_rpc_utils");
+  if (module_ == nullptr) {
+    throw std::runtime_error("import internal_rpc_utils module is NULL");
+    PyErr_Print();
```
good catch:)

xush6528 commented on torch/distributed/internal_rpc_utils.py:
```diff
@@ -0,0 +1,27 @@
+        python_udf = pickle.loads(pickled_python_udf)
+        result = python_udf.func(*python_udf.args, **python_udf.kwargs)
+    except Exception as e:
+        result = "run_python_udf_internal caught exception: " + str(e)
```
This is the line that silently swallows Exceptions

zhaojuanmao replied:
@xush6528 I add this exception in python land and pass it back to client in this way, because we do not have general remote exception handling yet. if we do not pass it in this way and there is function call exception here, server will crash. once we have general remote exception, we can have a new type of message and make it more clear to users. cc @mrshenli  does it make sense for error handling?

zhaojuanmao replied:
so right now users need to know how to validate result is expected or exception, before we add remote exception handling

mrshenli replied:
Yes, that makes sense to me. And I think it can come in a followup PR, right?

xush6528 replied:
@zhaojuanmao 
Thanks for clarifying the plan.

So, in recap, the plans is that,
we will have a unified error handling in C++ world that can handle
- built-in op error.
- python function error.

So leave this workaround here for now, and improve it later.

aazzolini commented:
What was the argument to changing the API from rpc_sync and rpc_async into "rpc" ?
1) this will make type-checking weaker
2) it's no clear why the default is sync
3) changing the argument will require changing the code that deals with the result type.

I don't see actually any argument for merging them into a single function.

mrshenli commented:
@aazzolini 

> What was the argument to changing the API from rpc_sync and rpc_async into "rpc" ?

The argument is `async_call=True/False`. 

> 1. this will make type-checking weaker

Could you please elaborate? And also for the 3rd bullet. Would it be efficient if we use `get_type_hints ` on UDFs and then use `type()` and `__class__` to get argument types?  

> 2. it's no clear why the default is sync

We were trying to make it consistent with c10d ops which all default to sync.

> I don't see actually any argument for merging them into a single function.

1. We were thinking about making it possible to add reserved kwargs in the future, e.g., `*args, timeout=5, **kwargs`.
2. The current form is also consistent with c10d ops, where `async_op` is an argument. 

If merging them together makes other things much harder to implement, we can revert it, but would like to know what are those pain points.

mrshenli commented:
I think test failure is my fault, but I don't know why it passed in other PRs (e.g., [1](https://circleci.com/gh/pytorch/pytorch/2393461?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)). Let me retest.

```
Aug 07 20:15:27   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/distributed/rpc.py", line 22, in _collect_worker_names
Aug 07 20:15:27     name_bytes = list(array.array('B', name_bytes))
Aug 07 20:15:27 TypeError: must be char, not unicode
```

aazzolini commented:
@mrshenli these are not specific pain-points, it's mostly for API design considerations.
The main issue here is that rpc(async=True/False) doesn't have a clear return type. It returns T if async is True and Future<T> if async is False. This weakens type checking. Think about how you'd do this in C++  -- you'd probably have 2 different functions with different result types.

I'm basically saying that we shouldn't be doing duck-typing here since it's so simple to just split this call into sync and async versions.

mrshenli commented:
@aazzolini I see. Any suggestion for reserving kwargs (e.g., timeout=5) in py2 and py3? 

BTW, I recall now that another reason for making the merge is that `sync_rpc` and `rpc_sync` collide a bit and might confuse people. And there is an ongoing discussion on whether we need `sync_rpc`, or should we just leave all synchronization efforts to applications.

xush6528 commented:
@mrshenli 

assume you made this the interace
```Python
def rpc(to, func, *args, timeout=5, **kwargs):
  pass
```

What if a client wants to call 
```Python
ask_server_to_do_some_heavy_work_but_only_within_a_timeout(timeout=10)
```

```Python
rpc(to="server", ask_server_to_do_some_heavy_work_but_only_within_a_timeout, timeout=10)
```
How this would work?

mrshenli commented:
@xush6528 

Good point! I don't have a solution for that. As @aazzolini's concern is mainly on the return type, maybe we can still keep `args=()` and `kwargs={}`, and just split `rpc` into different synchronous and asynchronous functions. Any thoughts on the name? I remember there was a complain that `rpc_sync` and `rpc_async` look too similar. What about `rpc` and `rpc_async`?

aazzolini commented:
@mrshenli , there's two discussions here. One is the input arguments, the other one is the result type. For the input aguments I do not have strong opinion.
However, for the result type, I do believe that each function should have a well defined result type. That's why we should have two separate functions for sync and async RPC.
If there's a discussion of not exposing synchronous RPC, then we should make asynchronous the default. So "dist.rpc()" would always be asynchronous. And if the user wants to make it synchronous we can always do "dist.rpc_sync()" on top.
Again my only point here is that we shouldn't have a function return type be conditioned by a boolean argument in the input.

mrshenli commented:
@zhaojuanmao if the plan is to only support this in PY3, let's make it explicit in docs and imports as well. I think `torch/distributed/__init__.py` should be fine as imports won't break (please do correct me if I am wrong, we definitely don't want to disable the entire distributed package just because of RPC), but we probably should error out when they call `init_rpc` in PY2. Does this make sense?

zdevito commented on torch/distributed/internal_rpc_utils.py:
```diff
@@ -0,0 +1,34 @@
+    return pickle.loads(pickled_python_result)
+
+
+PythonUDF = collections.namedtuple("PythonUDF", ["func", "args", "kwargs"])
```
I think we should make sure the data type sent over the wire to run an RPC is the same whether it is a torchscript or Python object. As far as I know, the torchscript pathway does not use a named tuple to transport this information. We also have to anticipate the future when Remote objects exists. What happens in the case where  we `torch.rpc(.., remote_object.foo, a, b)`?  would this still work if we are directly pickling the method of the object we are sending?

zhaojuanmao replied:
> As far as I know, the torchscript pathway does not use a named tuple to transport this information. 
how torchscript transport this information? 

> What happens in the case where we `torch.rpc(.., remote_object.foo, a, b)`? 
remote_object will be a class instance, right? unit tests showed pickle class instance method works. Are there some caveats I did not aware of?

zdevito reviewed (commented):
This looks pretty good, substantive changes:

1. we should be consistent about pybind vs no pybind usage, to prevent future bugs.
2. we should be consistent about the payload format across torchscript and python.

zhaojuanmao commented:
Thanks @zdevito! 

> 1. we should be consistent about pybind vs no pybind usage, to prevent future bugs.
   are you suggesting not to use python embedding API, instead it is better to use pybind APIs?
> 2. we should be consistent about the payload format across torchscript and python.
    why we need to make them consistent? I though torchscript will work like as builitin ops, we only need to get its qualified name and pass it over wire, then server can look up function based on the qualified name and run it in C++ land? looks like torchscript and python are executed in different code paths

zhaojuanmao commented:
> @zhaojuanmao if the plan is to only support this in PY3, let's make it explicit in docs and imports as well. I think `torch/distributed/__init__.py` should be fine as imports won't break (please do correct me if I am wrong, we definitely don't want to disable the entire distributed package just because of RPC), but we probably should error out when they call `init_rpc` in PY2. Does this make sense?

yeah, it is hard to resolve python2 issue. 
1)  python2 pickle can not pickle class static method and etc 
2) when parse binary string from C++ to pyObject, python2 and python3 need different format string

pritamdamania87 commented:
@mrshenli @zhaojuanmao 

The current API for rpc is: `def rpc(to, func, args=None, kwargs=None, async_call=False):`

Shouldn't we be using *args and **kwargs instead? Also, can we add some unit tests to test_rpc.py which use a mix of args and kwargs and also maybe only uses kwargs?

Regarding @xush6528's concern above (timeout=10 example), I think we have the same problem in `invoke_rpc_builtin` currently.

mrshenli commented:
@pritamdamania87 

> Regarding @xush6528's concern above (timeout=10 example), I think we have the same problem in `invoke_rpc_builtin` currently.

Yep, that's not a problem specific for this PR. And I think can be fixed in follow PRs. It's a good point to add tests for kwargs, and I think this can also be done in followup PRs?

zhaojuanmao commented on torch/csrc/distributed/rpc/PythonRpcHandler.h:
```diff
@@ -0,0 +1,28 @@
+namespace distributed {
+namespace rpc {
+
+class PythonRpcHandler {
```
I want to hold members into the class, but these members can be initialized only once and can only be accessed by its interface. Also class interfaces can be accessed everywhere. 
Does singleton class makes more sense?

xush6528 replied:
How does namespace + global function/variable sound?

zhaojuanmao commented:
> @mrshenli @zhaojuanmao
> 
> The current API for rpc is: `def rpc(to, func, args=None, kwargs=None, async_call=False):`
> 
> Shouldn't we be using *args and **kwargs instead? Also, can we add some unit tests to test_rpc.py which use a mix of args and kwargs and also maybe only uses kwargs?
> 
> Regarding @xush6528's concern above (timeout=10 example), I think we have the same problem in `invoke_rpc_builtin` currently.

@pritamdamania I can add some tests to use kwargs

mrshenli commented on torch/distributed/rpc.py:
```diff
@@ -156,17 +164,19 @@ def rpc(to, func, args=None, kwargs=None, async_call=False):
         >>> dist.init_rpc("worker1")
         >>> dist.join_rpc()
```
nit: shall we add an example for Python UDF?

mrshenli commented on torch/distributed/rpc.py:
```diff
@@ -73,13 +76,18 @@ def init_rpc(name, backend='pg'):
                        process group backend ``"pg"`` is the only available
                        backend implementation. (default: ``"pg"``).
     """
+    if sys.version_info < (3, 0):
```
nit: I saw `test_rpc.py` uses six. Shall we stay consistent in RPC-related files?

@ezyang do we have any preference on `six.PY2` vs `sys.version_info`?

ezyang replied:
PyTorch does NOT have a dependency on six in the main library (we have a shitty version of six in `torch._six`).  @soumith do you remember why we don't directly depend on six? It seems like a pretty benign dependency, as far as dependencies go.

mrshenli reviewed (commented):
@zhaojuanmao do you plan to address all of @zdevito's comments in this PR, or is there any of them can be made into followup ones?

ezyang commented on torch/distributed/rpc.py:
```diff
@@ -73,13 +76,18 @@ def init_rpc(name, backend='pg'):
     use ``pg`` (ProcessGroup) backend, ``init_process_group`` must be invoked
     prior to this method.
 
+    Also to be noted: RPC package only support Python3
```
This belongs at the top level module documentation, not here...

zhaojuanmao replied:
@ezyang we do not have rpc document in top level distributed module documentation yet. we will add this note when we have rpc document

ezyang commented on torch/distributed/rpc.py:
```diff
@@ -105,7 +113,7 @@ def rpc(to, func, args=None, kwargs=None, async_call=False):
     Arguments:
         to (str): name of the destination worker.
-        func (callable): a builtin function (e.g., ``torch.add``).
+        func (callable): a builtin function or user defined python function (e.g., ``torch.add``).
```
Another way to spell the docs here, is to just say any callable is fine, but then say that builtin functions (like `torch.add`) can be sent over RPC more efficiently.

ezyang commented on torch/distributed/rpc.py:
```diff
@@ -156,17 +164,19 @@ def rpc(to, func, args=None, kwargs=None, async_call=False):
         >>> dist.init_rpc("worker1")
         >>> dist.join_rpc()
     """
+    assert (callable(func)), "function should be callable."
```
Use of assert here is incorrect. An AssertionFailure means that we, the library authors, messed up. This is error checking in the public API; you should raise a `TypeError` here.

ezyang commented on torch/distributed/rpc.py:
```diff
@@ -15,8 +18,8 @@ def _collect_worker_names(name, group):
-    name_bytes = name if sys.version_info < (3, 0) else bytes(name, 'utf8')
-    name_bytes = list(array.array('B', name_bytes))
+    name_bytes = bytes(name) if sys.version_info < (3, 0) else bytes(name, 'utf8')
+    name_bytes = list(array.array(str('B'), name_bytes))
```
If Python 2 is not supported, why is there a version test here?

ezyang commented on test/test_rpc.py:
```diff
@@ -110,5 +134,74 @@ def test_join_rpc(self):
+        self.assertEqual(ret, my_class.my_static_method(n + 10))
+
+    @_wrap_with_rpc
+    def test_py_multi_aync_call(self):
```
nit: async

ezyang commented on test/test_rpc.py:
```diff
@@ -110,5 +134,74 @@ def test_join_rpc(self):
+    def test_py_function_exception(self):
+        n = self.rank + 1
+        dstRank = n % self.world_size
+        ret = dist.rpc('worker%d' % dstRank, no_result, args=(10,))
```
Please use `.format()`; the percent syntax is not recommended. (Unfortunately, as long as we support Python 3.5 you can't use Python 3 f-strings)

ezyang commented on test/test_rpc.py:
```diff
@@ -1,17 +1,40 @@
+
+    @staticmethod
+    def my_static_method(f):
+        return f > 10
```
nit: It would be clearer in the tests if the functions here also gave an indication of what they did (since you rely on the implementation details of these functions to compute what the correct test results are.)

ezyang commented on torch/csrc/distributed/rpc/PythonRpcHandler.h:
```diff
@@ -0,0 +1,34 @@
+private:
+  static PyObject* module_;
+  static PyObject* runUDFFunction_;
+  static PyObject* loadResultFunction_;
```
For new code we are writing, there's NO reason to use raw `PyObject*` pointers. The manual refcounting is an endless bugfarm. Please use `THPObjectPtr`, or you can also use the pybind11 equivalent if you like.

ezyang commented on torch/distributed/__init__.py:
```diff
@@ -16,3 +18,4 @@ def is_available():
     # this.
     from .distributed_c10d import _backend  # noqa: F401
     from .rpc import *  # noqa: F401
+    from .internal_rpc_utils import *  # noqa: F401
```
If rpc is Python 3 only, you really shouldn't import this (or `.rpc`) if you're on Python 2, right?

Also, why are we importing all of `internal_rpc_utils` into the PUBLIC module `torch.distributed`? That makes no sense to me. If they are internal, please access them through full module qualification; don't pollute public namespace just for the sake fo convenience.

ezyang commented on torch/csrc/distributed/rpc/PythonRpcHandler.cpp:
```diff
@@ -0,0 +1,96 @@
+void PythonRpcHandler::init() {
+  if (!Py_IsInitialized()) {
+    Py_Initialize();
+    pyInitializerCalled_ = true;
```
Under what circumstances will we be sending Python UDFs to remote processes that aren't already running Python? Does it simplify matters if you can assume that Python workers always start up from Python?

ezyang reviewed (changes requested):
Beyond some of the smaller requests, at a larger level I'd like some clarity on the lifetime of PythonRpcHandler object, and the relationship between C++ only execution versus C++ and Python execution.

Sorry about the late review; release was this week and I kept putting this PR review off XD

zhaojuanmao commented:
Thanks for your comments @ezyang! 

1) for PythonRpcHandler, we do not want it to have same lifetime as processGroupAgent, because some interface like "loadPythonUDFResult()" need to be called without processGroupAgent. Ideally all interfaces except init() of PythonRpcHandler should be able to be called everywhere.  I think it makes sense to make it as singleton class or just make it as namespace + global functions/variables.  How do you think?

2) "the relationship between C++ only execution versus C++ and Python execution." -- would you please clarify it more? I do not quite understand it :(.

built in op can be executed remotely in C++ purely, because JIT support so. 
for pure python function, we have to pickle it and unpickle it on python land, and execute it on python land as well. that is why C++ and python execution are mixed

zhaojuanmao commented:
Thanks everyone for giving such great feedback!

I resolved most of comments including switch c python API to pybind11 API, and others. 

Would like to address the comment about making binary format consistent btw python UDF and builtin operator in follow up PR.

mrshenli reviewed (approved):
Thanks for adding Python UDF support for RPC! 

Approving, but please consider performing the following two actions before merging:

1. Check with @gqchen to see if there is any additional blocking change requests. 
2. Make sure this works under OSX. I am not sure if CI tests for MaxOS here are sufficient. #23228 run into this problem previously.

ezyang commented on torch/CMakeLists.txt:
```diff
@@ -230,6 +230,7 @@ if (USE_DISTRIBUTED)
         ${TORCH_SRC_DIR}/csrc/distributed/rpc/ProcessGroupAgent.cpp
         ${TORCH_SRC_DIR}/csrc/distributed/rpc/functions.cpp
         ${TORCH_SRC_DIR}/csrc/distributed/rpc/python_functions.cpp
+        ${TORCH_SRC_DIR}/csrc/distributed/rpc/PythonRpcHandler.cpp
```
I'd aim for self-consistency, and the prevailing convention in torch/csrc is snake case.

pritamdamania87 commented:
> Yep, that's not a problem specific for this PR. And I think can be fixed in follow PRs. It's a good point to add tests for kwargs, and I think this can also be done in followup PRs?

Can we create github issues for things to address in followup PRs? This way we wouldn't lose track of these things.

mrshenli commented:
> Can we create github issues for things to address in followup PRs? This way we wouldn't lose track of these things.

Yep, see #24252 #24249  and #24247

zhaojuanmao commented:
the windows build failure is not related, all other checks are clean. macos build locally, import torch and they are fine

zhaojuanmao committed:
python udf over rpc (#23569)

Summary:
This diff is to support python user defined function over rpc for https://github.com/pytorch/pytorch/issues/23110, work flow is like this:
1. pickle python udf
2. pass pickle to C++
3. C++ pass over rpc from client to server
4. server call runPythonUDF() python function to unpickle and run python udf and pickle the udf result using python embedder
6. pass back serialized result from server to client
7. client call loadPythonUDFResult() python function to unpickle result
7. return it to python

right now, put rpc_sync_builtin() and rpc_async_builtin() as temporary interfaces for builtin operator remote calls, they accept qualified name string, this interface can execute builtin operators in C++ land.

rpc_sync() and rpc_async() accept python callables only right now, it could be user define python functions or builtin operator python functions, the python functions will be executed in python land.

once we can resolve builtin operator python callables to qualified name string, we can merge rpc_sync_builtin() into rpc_sync() then
Pull Request resolved: https://github.com/pytorch/pytorch/pull/23569

Test Plan: unit tests

Differential Revision: D16390764

fbshipit-source-id: 9373024280d08bc953391464f221e8ab65e9ba10
